chainerio/filesystems/hdfs.py

```diff
@@ -89,14 +94,39 @@ def info(self):
         info_str = "using hdfs"
         return info_str
 
-    def list(self, path_or_prefix: str = None):
-        if not path_or_prefix or None is path_or_prefix:
+    def list(self, path_or_prefix: str = None, recursive=False):
+        if path_or_prefix is None:
             path_or_prefix = "/user/{}".format(self.username)
 
         self._create_connection()
-        dir_list = self.connection.ls(path_or_prefix)
-        for _dir in dir_list:
-            yield os.path.basename(_dir)
+        target_dir = self.connection.info(path_or_prefix)
+        if "directory" != target_dir['kind']:
+            return None
+
+        target_path = target_dir['path'] + "/"
+        if not path_or_prefix.endswith("/"):
+            path_or_prefix = path_or_prefix + "/"
+
+        prefix_index = len(target_path[:-len(path_or_prefix)])
```

Comment:
Both the name prefix and index is confusing here, as it is rather "length" of a prefix, prefix of full path like `hdfs://...` and that is different from `path_or_prefix`. Should be named differently like `print_index`, `prefix_end_index`, or `full_path_prefix_len` .